libmount/src/context_mount.c

```diff
@@ -1068,6 +1138,88 @@ static int parse_ownership_mode(struct libmnt_context *cxt)
 	return 0;
 }
 
+#ifdef UL_HAVE_MOUNT_API
+/*
+ * Process X-mount.idmap=.
+ */
+static int mnt_context_prepare_idmaps(struct libmnt_context *cxt)
+{
+	int rc;
+	char *value;
+	size_t valsz;
+	char *saveptr = NULL, *tok;
+
+	const char *o = mnt_fs_get_user_options(cxt->fs);
+	if (!o)
+		return 0;
+
+	rc = mnt_optstr_get_option(o, "X-mount.idmap", &value, &valsz);
+	if (rc < 0)
+		return -MNT_ERR_MOUNTOPT;
+	if (rc > 0)
+		return 0;
+
+	if (!valsz)
+		return errno = EINVAL, -MNT_ERR_MOUNTOPT;
+
+	/* Has the user given us a path to a user namespace? */
+	if (startswith(value, "/")) {
+		cxt->userns_fd = open_userns(value);
+		if (cxt->userns_fd < 0)
+			return errno = EINVAL, -MNT_ERR_MOUNTOPT;
+
+		return 0;
+	}
+
+	for (tok = strtok_r(value, " ", &saveptr); tok;
+	     tok = strtok_r(NULL, " ", &saveptr)) {
+		char type;
+		struct id_map *idmap;
+		idmap_type_t map_type;
+		__u32 nsid = -1, hostid = -1, range = -1;
+
+		rc = sscanf(tok, "%c:%u:%u:%u", &type, &nsid, &hostid, &range);
+		if (rc != 4)
+			goto err;
+
+		switch (type) {
+		case 'g':
+			map_type = ID_TYPE_GID;
+			break;
+		case 'u':
+			map_type = ID_TYPE_UID;
+			break;
+		case 'b':
+			map_type = ID_TYPE_UIDGID;
+			break;
```

Comment:
maybe imply "b" prefix is not specified? i.e. if first char is a number, assume "b" is meant? I figure in 99% of the cases that's what people want

Reply:
Yeah, good idea.